src/thirdParty/spy.ts

```diff
@@ -0,0 +1,516 @@
+    stub.mockClear()
+    implementation = () => undefined as unknown as TReturns
+    onceImplementations = []
+    return stub
+  }
+
+  stub.mockRestore = () => {
+    stub.mockReset()
+    state.restore()
+    implementation = undefined
+    return stub
+  }
+
+  stub.getMockImplementation = () => implementation
+  stub.mockImplementation = (fn: (...args: TArgs) => TReturns) => {
+    implementation = fn
+    state.willCall(mockCall)
+    return stub
+  }
+
+  stub.mockImplementationOnce = (fn: (...args: TArgs) => TReturns) => {
+    onceImplementations.push(fn)
+    return stub
+  }
+
+  function withImplementation(fn: (...args: TArgs) => TReturns, cb: () => void): MockInstance<TArgs, TReturns>
+  function withImplementation(fn: (...args: TArgs) => TReturns, cb: () => Promise<void>): Promise<MockInstance<TArgs, TReturns>>
+  function withImplementation(fn: (...args: TArgs) => TReturns, cb: () => void | Promise<void>): MockInstance<TArgs, TReturns> | Promise<MockInstance<TArgs, TReturns>> {
+    const originalImplementation = implementation
+
+    implementation = fn
+    state.willCall(mockCall)
+    implementationChangedTemporarily = true
+
+    const reset = () => {
+      implementation = originalImplementation
+      implementationChangedTemporarily = false
+    }
+
+    const result = cb()
+
+    if (result instanceof Promise) {
+      return result.then(() => {
+        reset()
+        return stub
+      })
+    }
+
+    reset()
+
+    return stub
+  }
+
+  stub.withImplementation = withImplementation
+
+  stub.mockReturnThis = () =>
+    stub.mockImplementation(function (this: TReturns) {
+      return this
+    })
+
+  stub.mockReturnValue = (val: TReturns) => stub.mockImplementation(() => val)
+  stub.mockReturnValueOnce = (val: TReturns) => stub.mockImplementationOnce(() => val)
+
+  stub.mockResolvedValue = (val: Awaited<TReturns>) =>
+    stub.mockImplementation(() => Promise.resolve(val as TReturns))
+
+  stub.mockResolvedValueOnce = (val: Awaited<TReturns>) =>
+    stub.mockImplementationOnce(() => Promise.resolve(val as TReturns))
+
+  stub.mockRejectedValue = (val: unknown) =>
+    stub.mockImplementation(() => Promise.reject(val))
+
+  stub.mockRejectedValueOnce = (val: unknown) =>
+    stub.mockImplementationOnce(() => Promise.reject(val))
+
+  Object.defineProperty(stub, 'mock', {
+    get: () => mockContext,
+  })
+
+  state.willCall(mockCall)
+
+  mocks.add(stub)
+
+  return stub as any
+}
+
+export function fn<TArgs extends any[] = any, R = any>(): Mock<TArgs, R>
+export function fn<TArgs extends any[] = any[], R = any>(
+  implementation: (...args: TArgs) => R
+): Mock<TArgs, R>
+export function fn<TArgs extends any[] = any[], R = any>(
+  implementation?: (...args: TArgs) => R,
+): Mock<TArgs, R> {
+  const enhancedSpy = enhanceSpy(tinyspy.internalSpyOn({ spy: implementation || (() => {}) }, 'spy'))
+  if (implementation)
+    enhancedSpy.mockImplementation(implementation)
+
+  return enhancedSpy as Mock
+}
```

Comment:
Is this file copied from [`@vitest/spy`](https://github.com/vitest-dev/vitest/blob/main/packages/spy/src/index.ts)?  Could we just install that package?

Reply:
Hmmm, I see the issue, it needs bundling

Reply:
Got the bundle working for ESM, now looking at why [`@rollup/plugin-node-resolve`](https://github.com/rollup/plugins/tree/master/packages/node-resolve/#rollupplugin-node-resolve) is not resolving `@vitest/spy`...

Reply:
🙌

Reply:
`@vitest/spy` is using conditional exports without a `require` or `default`, fixing this means we can use it for CJS too.  I raised https://github.com/vitest-dev/vitest/pull/4462

Reply:
The above PR was merged and released with `@vitest/spy` v.1.0.0-beta.5, the rest of this work is now unblocked.

Reply:
nice 👌

Reply:
@christian-bromann I think we should release this as a major version bump with the old mocking and API access methods removed (all the APIs are now accessible in a much more fluid way with `execute`).  What do you think?

Reply:
The PR currently doesn't remove anything so it seems save to just do a patch with deprecations for the other commands. I am fine with either at the end. Let me know if I can help with the test. I want to present this next Thursday at a meetup.